libsignal-service-dotnet/websocket/WebSocketWrapper.cs

```diff
@@ -29,36 +29,96 @@ public class WebSocketWrapper
 
         public WebSocketWrapper(string uri, CancellationToken token)
         {
-            _ws = new ClientWebSocket();
-            _ws.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);
+            CreateSocket();
             _uri = new Uri(uri);
             Token = token;
         }
+        private void CreateSocket()
+        {
+            _ws = new ClientWebSocket();
+            _ws.Options.KeepAliveInterval = TimeSpan.FromSeconds(20);
+        }
 
         public void HandleOutgoingWS()
         {
             Debug.WriteLine(TAG + "HandleOutgoingWS started");
+            byte[] buf = null;
             while (!Token.IsCancellationRequested)
             {
+                
                 try
                 {
-                    var buf = OutgoingQueue.Take(Token);
+                    if (buf == null)
+                        buf = OutgoingQueue.Take(Token);
+                    ReconnectTask?.Wait();
                     _ws.SendAsync(new ArraySegment<byte>(buf, 0, buf.Length), WebSocketMessageType.Binary, true, Token).Wait();
+                    buf = null; //set to null so we do not retry the same block
                 }
                 catch (TaskCanceledException e)
                 {
+                    if (!Token.IsCancellationRequested)
+                        Reconnect();
                     Debug.WriteLine(TAG + "HandleOutgoingWS shutting down");
                 }
                 catch (Exception e)
                 {
                     Debug.WriteLine(TAG + "WS SendAsync failed: " + e.Message);
-                    //TODO reconnect
+                    Reconnect();
                 }
             }
             //TODO dispose
             Debug.WriteLine(TAG + "HandleOutgoingWS finished");
         }
+        private Task ReconnectTask;
+        public void Reconnect()
+        {
+            if (Token.IsCancellationRequested)
+                return;
+            if (ReconnectTask != null)
+            {
+                ReconnectTask?.Wait();//race condition handling
+                return;
+            }
+            var task = new TaskCompletionSource<bool>();
+            ReconnectTask = task.Task;
+            var tries = 0;
+            try
+            {
+                _ws.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, "goodbye", Token).Wait();
+            }
+            catch(Exception e)
+            {
+                Debug.WriteLine("exception while closing of: " + e.Message);
+            }
+            if (Token.IsCancellationRequested)
+                return;
+            while (true)
+            {
+                try
+                {
+                    tries++;
+                    CreateSocket();
```

Comment:
Do we need a new ClientWebSocket, or can we just call `ConnectAsync` on the old one?